db/crud_test.go

```diff
@@ -956,3 +956,211 @@ func BenchmarkHandleRevDelta(b *testing.B) {
+
+	// Create the third revision of the document
+	payload = `{"model":"1483C002SKU","sku":"5578523","name":"Canon EOS 5D Mark IV","price":"$2,499.99","version":"3a"}`
+	doc3, rev3, err := db.PutExistingRevWithBody("camera", unjson(payload), []string{"3-a", "2-a"}, false)
+	assert.NoError(t, err, "Couldn't create document")
+	assert.NotEmpty(t, doc3, "Document shouldn't be empty")
+	assert.Equal(t, "3-a", rev3, "Provided input revision ID should be returned")
+
+	// Get the 1x revision from third doc.
+	bodyBytes, removed, err = db.get1xRevFromDoc(doc3, rev3, true)
+	assert.NotEmpty(t, bodyBytes, "Document body bytes should be returned")
+	assert.False(t, removed, "This shouldn't be a removed document")
+	assert.NoError(t, response.Unmarshal(bodyBytes))
+	assert.Equal(t, "camera", response[BodyId])
+	assert.Equal(t, "3-a", response[BodyRev])
+	assert.Equal(t, "1483C002SKU", response["model"])
+	assert.Equal(t, "Canon EOS 5D Mark IV", response["name"])
+	assert.Equal(t, "$2,499.99", response["price"])
+	assert.Equal(t, "5578523", response["sku"])
+	assert.Equal(t, "3a", response["version"])
+
+	// Create the fourth revision by deleting the third revision
+	rev4, err := db.DeleteDoc("camera", rev3)
+	assert.NoError(t, err, "Document should be deleted")
+	assert.NotEmpty(t, rev4, "Document revision shouldn't be empty")
+
+	// Get the document body bytes with the deleted revision, list revisions enabled.
+	bodyBytes, removed, err = db.get1xRevFromDoc(doc3, rev3, true)
+	assert.NotEmpty(t, bodyBytes, "Document body bytes should be returned")
+	assert.False(t, removed, "This shouldn't be a removed document")
+	assert.NoError(t, response.Unmarshal(bodyBytes))
+	assert.Equal(t, "camera", response[BodyId])
+	assert.Equal(t, "3-a", response[BodyRev])
+	assert.Equal(t, "1483C002SKU", response["model"])
+	assert.Equal(t, "Canon EOS 5D Mark IV", response["name"])
+	assert.Equal(t, "$2,499.99", response["price"])
+	assert.Equal(t, "5578523", response["sku"])
+	assert.Equal(t, "3a", response["version"])
+}
+
+func TestCheckForUpgrade(t *testing.T) {
+	testBucket := base.GetTestBucket(t)
+	defer testBucket.Close()
+	bucket := testBucket.Bucket
+
+	context, err := NewDatabaseContext("db", bucket, false, DatabaseContextOptions{})
+	assert.NoError(t, err, "Couldn't create context for database 'db'")
+	defer context.Close()
+	db, err := CreateDatabase(context)
+	assert.NoError(t, err, "Couldn't create database 'db'")
+
+	// Create the first revision of the document
+	payload := `{"model":"1483C002SKU","sku":"5578523","name":"Canon EOS 5D Mark IV","price":"$2,499.99","version":"1a"}`
+	doc, rev1, err := db.PutExistingRevWithBody("camera", unjson(payload), []string{"1-a"}, false)
+	assert.NoError(t, err, "Couldn't create document")
+	assert.NotEmpty(t, doc, "Document shouldn't be empty")
+	assert.Equal(t, "1-a", rev1, "Provided input revision ID should be returned")
+
+	doc, buc := db.checkForUpgrade("camera", DocUnmarshalAll)
```

Comment:
This doesn't look like it handles all upgrade cases - are there already other tests for that?  

This also doesn't look like it handles the differences in behaviour when running with xattrs enabled/disabled.  When enabled, the PutExistingRevWithBody would write a doc w/ xattrs, and then not need to be upgraded.
When disabled, the checkForUpgrade is going to no-op.

Review other upgrade tests for examples of how to proceed.

Reply:
Yeah, changes were not included to handles all upgrade cases initially; but it has been added with the latest committed changes.